sdks/go/pkg/beam/core/runtime/exec/datasource.go

```diff
@@ -354,12 +354,12 @@ func (n *DataSource) Split(splits []int64, frac float64, bufSize int64) (SplitRe
 		return SplitResult{PI: s - 1, RI: s}, nil
 	}
 	// Otherwise, perform a sub-element split.
-	p, r, err := su.Split(fr)
+	ps, rs, err := su.Split(fr)
 	if err != nil {
 		return SplitResult{}, err
 	}
 
-	if p == nil || r == nil { // Unsuccessful split.
+	if len(ps) == 0 || len(rs) == 0 { // Unsuccessful split.
```

Comment:
Nit: nil seemed more explicit, why was this changed? (On that note, if one is empty must the other be empty as well, or is it OK to have one non-empty and treat that as an unsuccessful split?)

Reply:
Just because once I moved to slices, having an empty slice seemed to make more sense to indicate "the function worked, but the result was empty", rather than nil slices which is what gets returned for errors. I could switch it back to nil, since that was mostly a stylistic choice.

Checking over the code, yeah it looks like if one is empty I always just return both empty. But I think I used an "or" here just to avoid relying on implementation details of the split. Basically, if even one of them is missing then then Datasource knows the split failed, and doesn't have to really know what exactly happened.

Reply:
Idiomatically, if there's no intended semantic real difference between nil and empty, then checking the length is the correct way to handle that. nil slices are also 0 length. 
This catches the error case that empty slices are generated, and hints to the compiler that there's at least 1 entry in the slices after this if block.